docs/blog/posts/reworking-names.md

```diff
@@ -0,0 +1,49 @@
+---
+authors:
+  - lucas
+title: "Enhanced Representation and Management of Global Identifiers in hax"
+date: 2025-03-19
+---
+
+# Enhanced Representation and Management of Global Identifiers in hax
+
+In this technical overview, we present significant enhancements to the internal architecture used by the hax engine for representing and manipulating global identifiers. These updates address fundamental design limitations, greatly improving maintainability, clarity, and backend code generation precision.
+
+## Motivation for the Improvements
+
+Initially, hax assumed that all identifiers originated exclusively from Rust. While valid at the outset, this assumption quickly became inadequate. As hax evolved, new requirements emerged, prompting the engine to generate identifiers internally:
+
+- **Trait pre- and post-conditions:** in hax, these are explicitly represented as concrete methods within typeclasses. Conversely, in Rust, these conditions exist only as anonymous standalone functions.
+- **Explicit enum cast operations:** enum casts are primitive operations in Rust, but hax treats these casts as specialized operations, assigning distinct identifiers to them.
+- **Cross-module mutually recursive item bundles:** these bundles are internally introduced by hax, necessitating the generation of unique identifiers to prevent naming conflicts.
+
+Moreover, the previous identifier system lacked detailed metadata, such as the type of identifier (struct, function, type, etc.), complicating identifier rendering for backend tools.
+
+## Issues with the Previous Design
+
+Initially, identifiers were represented using slightly modified Rust `DefId`s accompanied by minimal metadata indicating the identifier's kind. This approach presumed that hax would never alter these `DefId`s but merely use those directly produced by the Rust compiler.
+
+This assumption was quickly challenged. The need to prefix or suffix identifiers emerged early, but the introduction of new internal modules completely disrupted the assumption. Identifiers had to be relocated across modules, representing a significant departure from the original design.
+
+As the API for manipulating identifiers grew increasingly permissive and transparent, the foundational assumption—that `DefId`s were unique, consistent, and Rust-generated—was entirely undermined. This resulted in numerous bugs in identifier rendering in backend outputs, leading to at least 16 documented issues ([#1135](https://github.com/cryspen/hax/issues/1135)).
+
+## Our New Approach
+
+The frontend has been enhanced to explicitly indicate the kind of each identifier, clarifying whether it represents a function, an associated type, a constant, etc. Additionally, it now provides detailed parent information, making the origin of identifiers more transparent. Alongside these improvements, we have redesigned our internal engine's identifier representation, introducing a layered structure where each layer addresses a distinct aspect.
+
```

Comment:
Show how the concrete example (problem) above was solved here?

Reply:
Ah, maybe I do have an example actually